### DISCUSSION TITLE
rqlite with a 2 node setup for an embedded application

### DISCUSSION THREAD

**Discussion by computinglife** [rqlite/rqlite#1806] [General]
Hi, 

Iam looking for a replicated SQLite database and seems like rqlite is exactly created for this. 

I do have some doubts - would be grateful if someone can answer them for me 

1. My cluster size has a minimum size of 2 and can be larger too - is there a way to get the clustering to work correctly in 2 nodes (Say by having the 2nd node as slave and then bump it up to participate in a write quorum once i have 3 nodes)?
2. Is there a way to compute how much bandwidth is used for the replication, especially during node restarts ? 
3. Is there a way to get commercial support ? 
4. Are there any cluster joining / restart order or such requirements - my application needs to be shipped to the customers and i want a static replication configured and setup between the nodes without any other customer intervention required later on. Hope Rqlite can be used for this scenario ? 

Thanks in advance
Arun

**Comment by otoolep** [comment]
Thanks for your questions. You should check out the docs on rqlite.io.

>My cluster size has a minimum size of 2 and can be larger too - is there a way to get the clustering to work correctly in 2 nodes (Say by having the 2nd node as slave and then bump it up to participate in a write quorum once i have 3 nodes)?

If one node is down in a 2-node cluster, then your cluster is down. See https://rqlite.io/docs/faq/#can-i-use-rqlite-to-replicate-my-sqlite-database-to-a-second-node. rqlite requires a majority of nodes to be up at all times. For a 2 node cluster, that is 2 nodes. You need a 3-node cluster to tolerate the failure of 1 node.

>Is there a way to compute how much bandwidth is used for the replication, especially during node restarts ?

Worse case (if the node that is restarted missed a substantial number of writes) the entire SQLite database will be transferred to the node that restarts. However this is one over a compressing connection.

>Is there a way to get commercial support ?

I do not offer any formal commercial support. What kind of support might you want?

>Are there any cluster joining / restart order or such requirements - my application needs to be shipped to the customers and i want a static replication configured and setup between the nodes without any other customer intervention required later on. Hope Rqlite can be used for this scenario ?

That is a somewhat vague question. I suggest you read the clustering guide and let me know if it doesn't answer your question. Generally speaking it's very easy to form clusters.

https://rqlite.io/docs/clustering/

**Reply by computinglife** [reply]
Thanks for responding. 

> For a 2 node cluster, that is 2 nodes. You need a 3-node cluster to tolerate the failure of 1 node.

I understand - i found what i needed in your docs - aka Read-Only nodes. I hope having a write-node + a read-only node would be sufficient to replicate the database to the 2nd node which is designated as read-only ?  

>Worse case (if the node that is restarted missed a substantial number of writes) the entire SQLite database will be transferred to the node that restarts. However this is one over a compressing connection.

What is substantial here ? is it configurable ? A 2 day maintenance window hopefully should not cause an entire DB to be transferred over WAN IMHO 

> I do not offer any formal commercial support. What kind of support might you want?

Oh just the sort where we could get 

1. breaking bugs with high impact fixed in a reasonable time 
2. Fixes for long term security issues 
3. help for emergency debugging - ( rarely, in case when we have an escalation which our support/engineers cannot root cause - hopefully around 2 or 3 an year tops)  

these are the only 3 scenarios i can think of.

**Comment by otoolep** [comment]
> I hope having a write-node + a read-only node would be sufficient to replicate the database to the 2nd node which is designated as read-only ?

Yes, that will probably work the way you want.

**Reply by computinglife** [reply]
Thanks - Can you also help me with how the cluster decides whether it needs to replicate the entire DB ? Is it based on how far behind in time the database is or based on the no of transactions its behind ? 

Also any guides on the type of SQL statements to avoid (eg probably not a good idea to use time in the SQL statements) ?

**Comment by otoolep** [comment]
https://rqlite.io/docs/api/non-deterministic/

https://rqlite.io/docs/api/api/#pragma-directives

**Comment by otoolep** [comment]
>Thanks - Can you also help me with how the cluster decides whether it needs to replicate the entire DB ?

It does it continually, streaming every request made to `/db/execute` to the read-only node.

**Comment by otoolep** [comment]
If a read-only node is down for a long time (such that the read-only node misses a few thousand updates, the exact number is configurable at the cost of disk space on the Leader) the Leader node may have to send the entire database to the read-only node when the read-only nodes comes back up and connects.

**Comment by otoolep** [comment]
Increase this value on the leader if you want, it would have the effect of keeping more data available for streaming to the read-only node, avoiding the need to send the whole database:

https://github.com/rqlite/rqlite/blob/master/cmd/rqlited/flags.go#L476

**Comment by otoolep** [comment]
>breaking bugs with high impact fixed in a reasonable time

I usually jump on any bugs that look serious. Most of the time they are fixed quite quickly.

>Fixes for long term security issues

Yes, are usually easy to fix because they mean just upgrading a dependency or underlying Docker image.

>help for emergency debugging - ( rarely, in case when we have an escalation which our support/engineers cannot root cause - hopefully around 2 or 3 an year tops)

I haven't done that kind of thing before (rqlite has a long history of very reliable performance so hasn't been deeded), but 
happy to learn more about what you have in mind. If you want to discuss more, book an Office Hours at https://rqlite.io/office-hours

**Comment by otoolep** [comment]
Office Hours is also good to discuss the issues you've outlined here.

**Comment by computinglife** [comment]
Thanks -  We will take you up on that offer. But first we need  to give it a spin to see how mature it is.
